src/backend/utils/misc/queryenvironment.c

```diff
@@ -978,8 +1111,469 @@ void ENRDropEntry(Oid id)
+						skip_cache_inval = true;
+				}
+
+				rel = relation_open(uncommitted_tup->catalog_oid, AccessExclusiveLock);
+
+				switch(uncommitted_tup->optype)
+				{
+					/* Peform the opposite operation as the saved optype. */
+					case ENR_OP_ADD:
+						_ENR_tuple_operation(rel, uncommitted_tup->tup, ENR_OP_DROP, skip_cache_inval);
+						break;
+					case ENR_OP_UPDATE:
+						ENRupdateTuple(rel, uncommitted_tup->tup);
+						break;
+					case ENR_OP_DROP:
+						_ENR_tuple_operation(rel, uncommitted_tup->tup, ENR_OP_ADD, skip_cache_inval);
+						break;
+				}
+				relation_close(rel, AccessExclusiveLock);
+			}
+		}
+
+		in_enr_rollback = false;
+
+		ENRDeleteUncommittedTupleData(InvalidSubTransactionId, enr);
+	}
+
+	/* Finally, clean up any ENRs that must be deleted. */
+	foreach(lc, enrs_to_drop)
+	{
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+		queryEnv->namedRelList = list_delete(queryEnv->namedRelList, enr);
+		pfree(enr->md.name);
+		pfree(enr);
+	}
+	MemoryContextSwitchTo(oldcxt);
+}
+
+/* This is used when rolling back a subtransaction. */
+/* It is similar to ENRRollback */
+void ENRRollbackSubtransaction(SubTransactionId subid, QueryEnvironment *queryEnv)
+{
+	ListCell 	*lc;
+	List *enrs_to_drop = NIL;
+	MemoryContext oldcxt;
+
+	if (!queryEnv)
+		return;
+
+	oldcxt = MemoryContextSwitchTo(queryEnv->memctx);
+
+	/* Add back any ENRs that must be restored. */
+	foreach (lc, queryEnv->dropped_namedRelList)
+	{
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+
+		Assert(enr->md.enrtype == ENR_TSQL_TEMP);
+		Assert(!enr->md.is_table_variable);
+
+		/* If this was dropped in the current subtransaction, then we must restore it. */
+		if (enr->md.dropped_subid != InvalidSubTransactionId && enr->md.dropped_subid >= subid) // 4
+		{
+			queryEnv->namedRelList = lappend(queryEnv->namedRelList, enr);		
+			queryEnv->dropped_namedRelList = foreach_delete_current(queryEnv->dropped_namedRelList, lc);
+			enr->md.dropped_subid = InvalidSubTransactionId;
+		}
+	}
+
+	/*
+	 * Loop through the registered ENRs to process each.
+	 */
+	foreach(lc, queryEnv->namedRelList)
+	{
+		Relation rel;
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+
+		if (enr->md.enrtype != ENR_TSQL_TEMP || enr->md.is_table_variable)
+			continue;
+
+		if (!enr->md.is_committed && enr->md.created_subid >= subid)
+		{
+			enrs_to_drop = lappend(enrs_to_drop, enr);
+			continue;
+		}
+
+		/* ROLLBACK block. */
+		in_enr_rollback = true;
+		for (int i = 0; i < ENR_CATTUP_END; i++)
+		{
+			List *uncommitted_cattups = enr->md.uncommitted_cattups[i];
+			ListCell *lc2;
+
+			foreach(lc2, uncommitted_cattups)
+			{
+				ENRUncommittedTuple uncommitted_tup = (ENRUncommittedTuple) lfirst(lc2);
+				bool skip_cache_inval = false;
+
+				if (!OidIsValid(uncommitted_tup->catalog_oid) || !HeapTupleIsValid(uncommitted_tup->tup))
+					elog(ERROR, "ROLLBACK failed! Invalid temp table entries were found.");	
```

Comment:
what happens when we encounter this error? a rollback cannot fail

Reply:
It should not fail, because this would mean a violated assumption or some issue with memory. But I thought it wouldn't be appropriate to emit a FATAL error to close the server since we are dealing with temp objects here - any bad state could be forcibly cleaned up by opening a new session.

Reply:
I think we will crash anyway.

For instance, try setting `uncommitted_tup->catalog_oid=0` then the code goes to error codepath over and over again until we hit a maximum.

ie
```
ereport(ERROR) -> triggers error handling
error handling goes to AbortTransaction
AbortTransaction calls ENRRollbackChanges
```

Reply:
Modified this to just continue - that way we can still attempt to clean up as much as possible and avoid infinite looping in error codepath.